Add support for scanning and querying indexes and some query operators

benhutchins opened the pull request:
This adds some basic support for scanning and querying indexes, helpful
when developing with indexes to see what data is being projected into
a index or to test queries.

It adds basic support for some query operators beyond the "is equal"
comparison, but this does not add rich operators like BETWEEN, begins
with, attribute exists, attribute does not exist, NULL, NOT, NULL, IN,
NOT IN, or the contains operators.

Renamed the "Scan" page to "Items", since Scan does not sound
appropriate given the new capabilities.

rchl commented on views/scan.ejs:
```diff
@@ -252,13 +300,13 @@
-            .attr('href', '?startKey=' + data.prevKey + '&pageNum=' + (data.pageNum-1) + '&'  + data.filterQueryString)
-          $('.page-item-previous').removeClass('disabled')
+            .attr('href', '?startKey=' + data.prevKey + '&pageNum=' + (data.pageNum-1) + '&' + data.filterQueryString)
+         $('.page-item-previous').removeClass('disabled')
```
One more indentation space here :)

rchl commented on views/scan.ejs:
```diff
@@ -252,13 +300,13 @@
-            .attr('href', '?startKey=' + data.nextKey + '&prevKey=' + data.prevKey + '&pageNum=' + (data.pageNum+1) + '&'  + data.filterQueryString)
+            .attr('href', '?startKey=' + data.nextKey + '&prevKey=' + data.prevKey + '&pageNum=' + (data.pageNum+1) + '&' + data.filterQueryString)
           $('.page-item-next').removeClass('disabled')
         }
```
A bit below please replace `showing items` with just `showing` as now `items` is redundant with `Items` in front of it already.

rchl commented on views/scan.ejs:
```diff
@@ -109,6 +109,31 @@
+            </th>
+            <td>
+              <select name="queryableSelection">
+                <option value="table"<% if (query.queryableSelection === 'table') { %> selected<% } %>>[Table] <%= Table.TableName %></option>
```
I feel that it's a bit redundant to show table name if there are no `GlobalSecondaryIndexes` since table name is shown already at the top of the page.

![Screenshot 2020-02-05 at 00 12 44](https://user-images.githubusercontent.com/153197/73796153-cc8d9800-47ac-11ea-94a4-2c9e3de7e35d.png)

(that dropdown doesn't have any other items)

benhutchins replied:
I am not sure if hiding it in this case makes a lot of sense though, I based some of this setup on the DynamoDB Console built into AWS Management Console.

If you want me to hide it I can do that change, for now, I've just disabled the select when there are no indexes available.

rchl commented on views/scan.ejs:
```diff
@@ -109,6 +109,31 @@
+            <th>
+              <select name="operationType">
+                <option value="scan"<% if (query.operationType === 'scan') { %> selected<% } %>>Scan</option>
+                <option value="query"<% if (query.operationType === 'query') { %> selected<% } %>>Query</option>
```
I've selected `query` operation on my table with no `GlobalSecondaryIndexes` and pressed "scan" and got error response. As an occasional user of DynamoDB, I don't exactly know how things should work and what's missing so it would be nice to have some error handling for this. For example, disable "scan" button if selected options make no sense. Or even disable `query` operation completely if possible to determine that it will not work.

benhutchins replied:
You can query your main index using the `KeySchema` fields. I've made the interface automatically add the required fields when switching to the `query` operation now so that should help prevent common user errors.

rchl commented on lib/backend.js:
```diff
@@ -377,36 +377,26 @@ exports.createServer = (dynamodb, docClient) => {
           filters,
-          pageNum: pageNum,
+          pageNum,
           filterQueryString: querystring.stringify(filters),
```
There is a bug now with handling next/previous page links. They don't preserve filters anymore.

```
> JSON.stringify(filters)
< "{"destination":{"operator":"=","value":"template"}}"

> querystring.stringify(filters)
< "destination="
```

benhutchins commented:
@rchl Thanks. I've gone through updating my pull request based on your feedback, please let me know if you find anything else.

rchl commented:
> @rchl Thanks. I've gone through updating my pull request based on your feedback, please let me know if you find anything else.

I'm still seeing quite a few UI quirks around the "query" method. I'm testing on a table without global secondary indexes.

 - Switching to "query" mode, the new filter is added with a pre-filled key but even if I fill out the value, I get an error on "search" (and an ugly alert one like "Error loading data: SyntaxError: Unexpected token < in JSON at position 0").
Testing on very simple DB like here:
<img width="605" alt="Screenshot 2020-02-05 at 11 52 08" src="https://user-images.githubusercontent.com/153197/73835643-08a51500-480e-11ea-913e-b288ecfb5dee.png">
 - Pressing search again turns the read-only filter into a normal filter (I guess it fails to get table structure at this point).
 - I think I'd prefer "Add Filter" button to be located within the "filters" frame. More relevant in context there IMO.

benhutchins commented:
@rchl Please review next round of updates, fixed the specific issues around tables without indexes, fixed the required flag going away, and moved the Add Filter button.

rchl commented:
> @rchl Please review next round of updates, fixed the specific issues around tables without indexes, fixed the required flag going away, and moved the Add Filter button.

Fixes looking good but found problem when pressing "Next" in paginated results.  Due to `&filters=undefined`, page crashes with `SyntaxError: Unexpected token u in JSON at position 0`.

rchl commented:
There are errors reported on trying to use `query` + a filter with non `=` condition. For any other condition I'm getting either `[ValidationException] Query key condition not supported` or `[ValidationException] Invalid operator used in KeyConditionExpression: <>`.

Is it possible to error-handle that better so those are, for example, not enabled?

BTW. I have added a bit better error reporting on errors triggered during items fetching at https://github.com/aaronshaf/dynamodb-admin/pull/91 . The error messages I've mentioned above are reported to the UI with that PR applied.

benhutchins commented:
@rchl It now disabled the operators that are not available based on the key type (HASH only allows equals, but RANGE allows for sort-based operators like greater or less than).

I also fixed the `filters=undefined` from getting into the query.

rchl commented on views/scan.ejs:
```diff
@@ -119,13 +119,19 @@
+                  <% Table.KeySchema.forEach(keySchemaItem => { %>
+                    data-key-<%= keySchemaItem.KeyType.toLowerCase() %>="<%= keySchemaItem.AttributeName %>"
+                  <% }) %>
+                  <% if (sortKey) { %> data-sort-key="<%= sortKey.AttributeName %>" <% } %>
```
`sortKey is not defined`

benhutchins replied:
Sorry about that, that was dead code, deleted that line

rchl commented:
Sorry but I have found one more problem. The condition type is properly restricted now initially but once you press `Search` it becomes editable again.

benhutchins commented:
@rchl You looking for a job as a QA/QE, even part-time?

Let's try this again. I've reworked the way filters are restored from the query string entirely so the required and available operators logic is only in once place still.

rchl commented:
> @rchl You looking for a job as a QA/QE, even part-time?

No thanks, I have job as developer already. :)
But I was working as a QA at the beginning of my career. :)

rchl commented:
I've skimmed a bit over the latest changes but couldn't find any problems while testing. I'll merge it tomorrow if nothing else comes up.

Thanks for really nice work.

benhutchins commented:
@rchl I found an issue, when you try to filter using operators like less/great than for non-KeySchema defined attributes, it does not work because it doesn't convert the value from a string into a number.

I don't think forcing all numeric strings to a number would be the smartest choice, because that would cause other race conditions.

I'm working on a fix now.

benhutchins commented:
@rchl fixed that.

I'll likely end up adding the advanced filtering logic (between, exists, is null, is not null, starts_with) at some point. I also think it might be useful to show the query that was run, it'd make the tool helpful for people looking into how to structure their search param objects. For another PR.

rchl commented on views/scan.ejs:
```diff
@@ -110,61 +110,73 @@
                         list="attributes"
                       />
                     </td>
                     <td>
```
Can we specify some "big enough" width for attribute type and operator columns? They stretch too much and there is no need since content is static.

benhutchins replied:
Done

rchl commented:
Thank you!
